### ISSUE TITLE
Automatic code hinting, without hitting tab

### ISSUE THREAD

**Issue by rkpatel33** [ipython-contrib/IPython-notebook-extensions#563]
Is it possible to turn on autocomplete so that the suggestion list is displayed on every letter key press in edit mode?

This would bring the typing experience closer to that of Atom or Sublime code complete and would make switching back and forth just a bit easier.

Would love to see this if it does not exist, or please let me know if someone knows an easy way to hack it in `custom.js`.

**Comment by jcb91** [comment]
I'm not aware of an existing implementation, but this does sound like a nifty idea, and should certainly be possible. I'll have a quick think.

**Comment by rkpatel33** [comment]
Cool. I had suggested in the main Jupyter repo issues to just connect the even attached to <tab> in edit mode to and edit mode keypress of [A-Zaz], but the fear was it was gum up the front end and introduce latency.

I'n not familiar enough with the code to have a feel for it.

**Comment by jcb91** [comment]
Indeed, it could end up making things a bit unresponsive, particularly where rounds trips to the kernel are slow (think kernels on remote machines, etc), so I can understand not wanting to implement it in core. This is in many ways a good thing for an extension, as people are free to enable/disable it as they choose...

**Comment by jcb91** [comment]
ok, I've made a basic version (in #564). It turned out to be a little more complicated than I'd anticipated - the code completer in notebook javascript is doing some interruption of the codemirror events, and has a start/modify/stop model, so it takes a little more handling. I think it's passable now, but it relies a little on a timeout, currently set to 200ms, which may need adjusting. If you could let me know if/how well it suits you, that'd be great...

**Comment by rkpatel33** [comment]
Awesome, I will give it shot. I installed Jupyter with the pip command. Is there a way I can switch to this branch, or do I need to pull the repo and run from there?

**Comment by jcb91** [comment]
So, since this a notebook extension, installing it should be relatively independent of your jupyter installation. The easiest route is to install the (`IPython-notebook-extensions`) repo using its `setup.py` script, (which isn't yet a full setuptools installer, but will modify the necessary config files, and copy things into the correct directories for you). This should work via pip, with

```bash
pip install https://github.com/jcb91/IPython-notebook-extensions/archive/hint_constantly.zip
```

Some people have had trouble with using pip directly, if it doesn't work for you, you can clone the repo, switch to the correct branch, and run the setup script directly:

```bash
git clone https://github.com/jcb91/IPython-notebook-extensions.git jcb91
cd jcb91
git checkout hint_constantly
python setup.py install
```

Once you've installed the repo, you'll need to restart your jupyter server, then at the url `/nbextensions`, you should find the configuration page, form which you can enable the notebook extension so that it'll load automatically for any notebook. 

Wow, that turned out rather longer than I expected

**Comment by jcb91** [comment]
Assuming this got fixed by #564
